### PR TITLE
fix: enforce content type header for API requests

### LINKED ISSUE #2496: Argo does not provide CSRF protection
**Describe the bug**
Currently, argocd does not provide CSRF protection. When users log in the UI, a malicious script could be executed on its behalf

**To Reproduce**
1. Expose argocd through load-balancer
2. Create an application
3. Paste the following in html file
```
<HTML>

    <HEAD>
    
    
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"></script>
    </HEAD>
    
    <BODY BGCOLOR="FFFFFF">
    
        <script>
            $.ajax({
                type: "DELETE",
                url: "https://<your-load-balancer-ip>/api/v1/applications/<application name from step 2>",
                data: "cascade=true",
                success: function(msg){
                    alert("Data Deleted: " + msg);
                }
            });
        </script> 
    </BODY>
    
    </HTML>
```
4. Start chrome without cors in order to test
```
https://alfilatov.com/posts/run-chrome-without-cors/
```
5. Log in to argocd UI
6. Open in new tab the created html file from step 3
7. Check the result in argocd UI

**Expected behavior**

This request should not be successful - 403 should be returned

**Version**
Any argocd version

### PR DESCRIPTION
**Pull request by alexmt**
Closes https://github.com/argoproj/argo-cd/issues/2496

PR implements content type enforcement for REST API calls. API server gets a new CLI flag `api-content-types` - a semi-column delimited list of content types. If list of content types is not empty then  any non GET REST API call must include content type header with one of the supported values. By default `application/json` is required. User might configure an empty list to disable content type enforcement.

### CODE REVIEW

**Review by gdsoumya — APPROVED** [review]
LGTM

**Review by jessesuen — APPROVED** [review]

**Comment by crenshaw-dev** [comment]
@alexmt can you take a look at the e2e failures?

**Comment by alexmt** [comment]
Oh, right. Updated e2e that patches instance using REST API call

File: cmd/argocd-server/commands/argocd_server.go [hunk 1]
```diff
@@ -240,6 +243,7 @@ func NewCommand() *cobra.Command {
 	command.Flags().StringVar(&repoServerAddress, "repo-server", env.StringFromEnv("ARGOCD_SERVER_REPO_SERVER", common.DefaultRepoServerAddr), "Repo server address")
 	command.Flags().StringVar(&dexServerAddress, "dex-server", env.StringFromEnv("ARGOCD_SERVER_DEX_SERVER", common.DefaultDexServerAddr), "Dex server address")
 	command.Flags().BoolVar(&disableAuth, "disable-auth", env.ParseBoolFromEnv("ARGOCD_SERVER_DISABLE_AUTH", false), "Disable client authentication")
+	command.Flags().StringVar(&contentTypes, "api-content-types", "application/json", "Semicolon separated list of allowed content types for non GET api requests. Any content type is allowed if empty.")
```
Comment (Nibiii) [inline cmd/argocd-server/commands/argocd_server.go:246]: Wouldn't this default value make disabling enforcing content type header impossible?

**Pull request merged**